PacketTunnelProvider/NetworkProtectionPacketTunnelProvider.swift

```diff
@@ -20,24 +20,142 @@
+            case .keychainDeleteError(let status): // TODO: Check whether field needed here
+                pixelEvent = .networkProtectionKeychainDeleteError
+                params[PixelParameters.keychainErrorCode] = String(status)
+            case .noAuthTokenFound:
+                pixelEvent = .networkProtectionNoAuthTokenFoundError
+            case .unhandledError(function: let function, line: let line, error: let error):
+                pixelEvent = .networkProtectionUnhandledError
+                params[PixelParameters.function] = function
+                params[PixelParameters.line] = String(line)
+                pixelError = error
+            }
+            DailyPixel.fireDailyAndCount(pixel: pixelEvent, error: pixelError, withAdditionalParameters: params)
+        }
+    }
+
+    public override func startTunnel(options: [String: NSObject]?, completionHandler: @escaping (Error?) -> Void) {
+        super.startTunnel(options: options) { error in
+            if error != nil {
+                DailyPixel.fireDailyAndCount(pixel: .networkProtectionFailedToStartTunnel, error: error)
+            }
+            completionHandler(error)
+        }
+    }
+
     @objc init() {
         let tokenStore = NetworkProtectionKeychainTokenStore(keychainType: .dataProtection(.unspecified),
                                                              errorEvents: nil)
-
+        let errorStore = NetworkProtectionTunnelErrorStore()
         super.init(notificationsPresenter: DefaultNotificationPresenter(),
                    tunnelHealthStore: NetworkProtectionTunnelHealthStore(),
-                   controllerErrorStore: NetworkProtectionTunnelErrorStore(),
+                   controllerErrorStore: errorStore,
                    keychainType: .dataProtection(.unspecified),
                    tokenStore: tokenStore,
-                   debugEvents: nil,
+                   debugEvents: Self.networkProtectionDebugEvents(controllerErrorStore: errorStore),
                    providerEvents: Self.packetTunnelProviderEvents)
+        startMonitoringMemoryPressureEvents()
+    }
+
+    private func startMonitoringMemoryPressureEvents() {
+        let source = DispatchSource.makeMemoryPressureSource(eventMask: .all, queue: nil)
+
+        let queue = DispatchQueue.init(label: "com.duckduckgo.mobile.ios.alpha.NetworkExtension.memoryPressure")
+        queue.async {
+            source.setEventHandler {
+                let event: DispatchSource.MemoryPressureEvent  = source.mask
+                print(event)
+                switch event {
+                case DispatchSource.MemoryPressureEvent.normal:
+                    break
+                case DispatchSource.MemoryPressureEvent.warning:
+                    Pixel.fire(pixel: .networkProtectionMemoryWarning)
+                case DispatchSource.MemoryPressureEvent.critical:
+                    Pixel.fire(pixel: .networkProtectionMemoryCritical)
```

Comment:
I'll be curious to see if these ones ever come through - will the extension have enough time to send the pixel before it gets killed? If we start getting warning pixels but never critical ones then that would be my theory, and we can look at implementing an offline pixel cache to catch situations like this.